snakemake/jobs.py

```diff
@@ -1192,24 +1218,79 @@ def merge(self, other):
+            pipe_dependencies = {}
+            for name, group in pipe_groups.items():
+                pipe_dependencies[name] = set(
+                    d
+                    for job in group
+                    for d in self.dag.dependencies[job]
+                    if d not in group
+                )
 
-    @property
-    def all_products(self):
-        if self._all_products is None:
-            self._all_products = set(f for job in self.jobs for f in job.products)
-        return self._all_products
+            # Collect every job's dependencies into a definitive mapping. Regular jobs
+            # get their dependencies directly from dag.dependencies, while pipe jobs
+            # get their dependencies from the pipe_dependencies calculated above.
+            dependencies = {}
+            for job in self.jobs:
+                if job.pipe_group in pipe_dependencies:
+                    dependencies[job] = pipe_dependencies[job.pipe_group]
+                else:
+                    dependencies[job] = self.dag.dependencies[job]
+
+            self.toposorted = [*self.toposort(self.jobs, dependencies)]
+
+    def toposort(self, group, dependencies):
+        """Sort group of jobs into layers of dependency.
+
+        Args:
+            group (Iterable[jobs]): List of jobs to sort
+            dependencies (Dict[job, Iterable[jobs]]): Dictionary of job keys mapped to
+                a list of all jobs they depend on. Dependencies should have one key
+                entry for every job in group.
+        """
+        from toposort import toposort as _toposort
+
+        # _toposort takes a dict: {job: {dependencies...}, ...}. Here, just filter
+        # out all the dependencies that aren't in the group.
+        yield from _toposort(
+            {job: {dep for dep in dependencies[job] if dep in group} for job in group}
+        )
 
     def __iter__(self):
         if self.toposorted is None:
             yield from self.jobs
-        else:
-            yield from chain.from_iterable(self.toposorted)
+            return
+
+        # The scheduler currently relies on Pipe jobs being returned in dependency order
+        # (i.e. if a -> b, yield a before b). finalize(), which produces self.toposort,
+        # puts all pipe jobs at the same toposort level for resource calculation
+        # purposes, so if we yield directly from toposort, we can't guarentee the order.
+        # Here, we filter through jobs to find pipe groups, then rerun toposort on each
+        # pipe group.
+        pipe_groups = defaultdict(set)
+        for sibling in self.toposorted:
+            for job in sibling:
+                if job.pipe_group is None:
+                    yield job
+                    continue
+                pipe_groups[job.pipe_group].add(job)
+
+            for group in pipe_groups.values():
+                yield from chain(*self.toposort(group, self.dag.dependencies))
 
```

Comment:
I think this should rather be done in finalize by resorting each level in the toposorted that is stored there with the strategy you use here. This is be better from a performance perspective.

Reply:
Good call. I actually didn't realize until recently that the `toposort` implementation was moved to the `dag` module, but I still had mine in `jobs`. I'll have a commit soon that unifies all the code under the `dag` version

Reply:
DAG has toposort as well, indeed. But it happens on a global scale there, not within a group. Nevertheless, it might indeed be a good idea to unify the two. Maybe put both into a new module under snakemake.common.

Reply:
I've update jobs so that it call the dag toposort in the same way it currently does on the master branch. If you prefer I move it, I can do that too